base/iterators.jl

```diff
@@ -1170,6 +1171,10 @@ IteratorEltype(::Type{Flatten{I}}) where {I} = _flatteneltype(I, IteratorEltype(
 IteratorEltype(::Type{Flatten{Tuple{}}}) = IteratorEltype(Tuple{})
 _flatteneltype(I, ::HasEltype) = IteratorEltype(eltype(I))
 _flatteneltype(I, et) = EltypeUnknown()
+# More refined handling of eltype and IteratorEltype for heterogeneous inner iterators,
+# when the outer iterator is a tuple.
+eltype(::Type{Flatten{I}}) where {I <: Tuple} = Base.typejoin(map(eltype, fieldtypes(I))...)
```

Comment:
```suggestion
eltype(::Type{Flatten{I}}) where {I <: Tuple} = eltype(map(eltype, fieldtypes(I)))
```
would this be better?

Reply:
Wouldn't this always return `DataType`?

Reply:
oh right, it would have to be converted back to a Type first
```suggestion
eltype(::Type{Flatten{I}}) where {I <: Tuple} = eltype(Tuple{map(eltype, fieldtypes(I))...})
```